portal-bridge/src/census/mod.rs

```diff
@@ -42,91 +31,87 @@ pub const ENR_OFFER_LIMIT: usize = 4;
-                }
-                // yield next known state peer and ping for liveness
-                Some(Ok(known_enr)) = self.state.peers.next() => {
-                    self.state.process_enr(known_enr.1.0).await;
-                    info!("Updated state census: found peers: {}", self.state.peers.len());
-                }
-                Some(Ok(known_enr)) = self.beacon.peers.next() => {
-                    self.beacon.process_enr(known_enr.1.0).await;
-                    info!("Updated beacon census: found peers: {}", self.beacon.peers.len());
-                }
+    /// Initialize subnetworks and starts background service that will keep our view of the network
+    /// up to date.
+    ///
+    /// Returns JoinHandle of the background service.
+    pub async fn init(
+        &mut self,
+        subnetworks: impl IntoIterator<Item = Subnetwork>,
+    ) -> Result<JoinHandle<()>, CensusError> {
+        if self.initialized {
+            return Err(CensusError::AlreadyInitialized);
+        }
+        self.initialized = true;
+
+        let subnetworks = HashSet::from_iter(subnetworks);
+        for subnetwork in &subnetworks {
+            info!("Initializing {subnetwork} subnetwork");
+            match subnetwork {
+                Subnetwork::History => self.history.init().await?,
+                Subnetwork::State => self.state.init().await?,
+                Subnetwork::Beacon => self.beacon.init().await?,
+                _ => return Err(CensusError::UnsupportedSubnetwork(*subnetwork)),
             }
         }
+
+        Ok(self.start_background_service(subnetworks))
     }
 
-    pub fn get_interested_enrs(&self, content_key: ContentKey) -> Result<Vec<Enr>, CensusError> {
-        match content_key {
-            ContentKey::History(content_key) => {
-                self.history.get_interested_enrs(content_key.content_id())
-            }
-            ContentKey::State(content_key) => {
-                self.state.get_interested_enrs(content_key.content_id())
-            }
-            ContentKey::Beacon(content_key) => {
-                self.beacon.get_interested_enrs(content_key.content_id())
+    /// Starts background service that is responsible for keeping view of the network up to date.
+    ///
+    /// Selects available tasks and runs them. Tasks are provided by enabled subnetworks.
+    fn start_background_service(&self, subnetworks: HashSet<Subnetwork>) -> JoinHandle<()> {
+        let mut history_network = self.history.clone();
+        let mut state_network = self.state.clone();
+        let mut beacon_network = self.beacon.clone();
+        let service = async move {
+            loop {
+                tokio::select! {
+                    peer = history_network.peer_to_process(), if subnetworks.contains(&Subnetwork::History) => {
+                        history_network.process_peer(peer).await;
+                    }
```

Comment:
I think there might be misunderstanding.

The `if` is not part of the regular code flow, but the special syntax for `tokio::select!` (more details [here](https://docs.rs/tokio/latest/tokio/macro.select.html)), and there isn't an option for `else`.
Unless you were thinking about `else` case of the `tokio::select!` statement, but then the comment doesn't make sense because code would already panic if `else` is not present and code execution would go there.
With the current implementation, this can happen if no subnetworks were provided in the `init` function, and now I added a check for that as well.

The `if` statements are needed with the current implementation because `peer_to_process()` would return `None` for uninitialized networks, forcing them to re-initialize and we don't want that for non-enabled subnetworks.

The original implementation was using selectors like this:
```
Some(Ok(known_enr)) = self.history.peers.next() => { ... }
```
(basically ignoring those `None` values) and re-initialization would be triggered when `get_interested_enrs()` would fail, causing that request to take much longer but still return empty set of enrs.

I believe that re-initialization should be detected and done as part of the regular census maintenance (and not be triggered by `get_interested_enrs`), and this is the cleanest way I found to do it.

I'm not sure if my explanation is clear. I think good understanding of what `tokio::select!` does under the hood is very important here, and I'm not sure how well I was able to explain my reasoning around it.